### PR TITLE
FIX: pass renderer through adjust_bbox

### LINKED ISSUE #27763: [Bug]: colorbar doesn't register inset_axis as cax
### Bug summary

When constructing inset axes using the tutorial on [this page](https://matplotlib.org/stable/gallery/axes_grid1/demo_colorbar_with_inset_locator.html) the gradient of the colorbar does not appear in the constructed inset axis but somewhere else on the figure (location seems to depend on figure size) once saved to a PDF with multiple other figures if bbox_inches='tight'. Similar to #27143 and #25176.

### Code for reproduction

```python
from mpl_toolkits.axes_grid1.inset_locator import inset_axes
import matplotlib.pylab as plt
from matplotlib.backends.backend_pdf import PdfPages

#make the first figure
fig, axes = plt.subplots(1, 1, facecolor = 'white', figsize = (20, 5), constrained_layout = True)

x = random.randint(100, size=(100, 2048))
c = axes.imshow(x, origin = 'lower', cmap = 'Reds', vmin = 0, vmax = 100, aspect = "auto")
axins1 = inset_axes(axes,
                    width="10%",
                    height="30%",
                    loc="upper left",
                    bbox_to_anchor = (0.045, 0., 1, 1),
                    bbox_transform = axes.transAxes
                    )
axins1.xaxis.set_ticks_position("bottom")
fig.colorbar(c, cax=axins1, orientation="horizontal")
axes.set_yticklabels([])

fig.suptitle("Title")
plt.figtext(-0.02, 0.5, 'Random', c = 'k', rotation = 90, ma = 'center', fontsize = medium)

#make figure 2
fig2, axes = plt.subplots(1, 1, facecolor = 'white', figsize = (20, 5), constrained_layout = True)

x = random.randint(100, size=(100, 2048))
c = axes.imshow(x, origin = 'lower', cmap = 'Reds', vmin = 0, vmax = 100, aspect = "auto")
axins1 = inset_axes(axes,
                    width="10%",
                    height="30%",
                    loc="upper left",
                    bbox_to_anchor = (0.045, 0., 1, 1),
                    bbox_transform = axes.transAxes
                    )
axins1.xaxis.set_ticks_position("bottom")
fig2.colorbar(c, cax=axins1, orientation="horizontal")
axes.set_yticklabels([])

fig2.suptitle("Title")
plt.figtext(-0.02, 0.5, 'Random', c = 'k', rotation = 90, ma = 'center', fontsize = medium)

#save figs to PDF
figs = [fig, fig2]
with PdfPages('test.pdf') as pdf:
    for fig in figs:
        pdf.savefig(fig, bbox_inches='tight')
```

### Actual outcome

The gradient for the colorbar appears outside of the inset axes on each of the plots
[test_small.pdf](https://github.com/matplotlib/matplotlib/files/14215554/test_small.pdf)

### Expected outcome

The result for this particular example when bbox_inches='tight' is not called in savefig(). Note: when running in jupyter lab this is always the outcome--only when saving the figure does the gradient jump out of the inset axis.
[test_loose.pdf](https://github.com/matplotlib/matplotlib/files/14215555/test_loose.pdf)

### Additional information

I tried the same suggestion as in #27143 and used `axins1 = axes.inset_axes([0.05, 0.7, 0.1, 0.2], transform = axes.transAxes)` in my own code which seemed to fix this. Unsure why the example constructor for inset_axes [here](https://matplotlib.org/stable/gallery/axes_grid1/demo_colorbar_with_inset_locator.html) doesn't work in this case.

### Operating system

Mac OS 13.6.3 (Ventura) Intel

### Matplotlib Version

3.8.0

### Matplotlib Backend

Installed osx event loop hook. MacOSX

### Python version

Python 3.9.18

### Jupyter version

Jupyter Lab 3.6.3

### Installation

conda

### PR DESCRIPTION
**Pull request by rcomer**
## PR summary

Closes #27763.

I have not fully understood what was going wrong, but I did discover that when [fig._get_renderer](https://github.com/matplotlib/matplotlib/blob/f8900ead0d9381a7652568768b065324f929734e/lib/mpl_toolkits/axes_grid1/inset_locator.py#L30) is called, `fig.dpi` gets reset to 72.  As the renderer had deliberately set it to something else [here](https://github.com/matplotlib/matplotlib/blob/f8900ead0d9381a7652568768b065324f929734e/lib/matplotlib/backends/backend_mixed.py#L77), this is probably not helpful!  Anyway, it is intuitive to me that the renderer used to work out where things are should be the same renderer that does this part of the drawing.  Taking a simplified version of the #27763 example,

```python
from mpl_toolkits.axes_grid1.inset_locator import inset_axes
import matplotlib.pyplot as plt
import numpy as np

fig, axes = plt.subplots(1, 1, dpi=100)

x = np.random.randint(100, size=(10, 200))
c = axes.imshow(x, origin='lower', cmap='Reds', aspect='auto')
axins1 = inset_axes(axes,
                    width="10%",
                    height="30%",
                    loc="upper left",
                    bbox_to_anchor = (0.045, 0., 1, 1),
                    bbox_transform = axes.transAxes
                    )

fig.colorbar(c, cax=axins1, orientation="horizontal")

fig.savefig('test_single.pdf', bbox_inches='tight')
```
With main:
image

With my branch:
image

The alignment is still not perfect (as also shown in the new test image), but it looks identical to what I get with v3.7.1 (before I started breaking things :sheep:).  I note that several commenters on the issue seemed happy with v3.7.1 behaviour.  Also I cannot see any offset if I increase the dpi to 300:

image

I propose to open a new issue to track the remaining small offset.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -2169,7 +2172,7 @@ def print_figure(
 
                 # call adjust_bbox to save only the given area
                 restore_bbox = _tight_bbox.adjust_bbox(
-                    self.figure, bbox_inches, self.figure.canvas.fixed_dpi)
+                    self.figure, bbox_inches, renderer, self.figure.canvas.fixed_dpi)
```
Comment (rcomer) [inline lib/matplotlib/backend_bases.py:2175]: Passing the renderer here does not affect the target issue but I assume that passing it is more efficient than potentially calling `_get_renderer` again.

**Review by timhoffm — APPROVED** [review]

**Review by jklymak — APPROVED** [review]
Is there a test we can do with the main library, or are the effects too subtle there?

**Comment by rcomer** [comment]
> Is there a test we can do with the main library, or are the effects too subtle there?

We need an axes that uses a locator that requires a renderer.  Do we have one of those in the main library?  I checked that `Axes.inset_axes`' locator works without a renderer.

Also I should have mentioned that this change does nothing for your example at https://github.com/matplotlib/matplotlib/issues/27763#issuecomment-1936312401

**Comment by jklymak** [comment]
OK< so there is still a small registration issue.  Thanks for checking.

**Comment by rcomer** [comment]
> I propose to open a new issue to track the remaining small offset.

#29383

**Pull request merged**